### PR TITLE
Refactoring some of the P2P breakout.

### PR DESCRIPTION
**Pull request by micahriggan**

### CODE REVIEW

File: packages/bitcore-node/src/services/p2p/index.ts [hunk 1]
```diff
@@ -16,6 +16,12 @@ import { CSP } from '../../types/namespaces/ChainStateProvider';
 import { setImmediate } from 'timers';
 import { EventEmitter } from 'events';
 
+let P2PClasses: {
```
Comment (illegalprime) [inline packages/bitcore-node/src/services/p2p/index.ts]: should be `const`

Comment (illegalprime) [inline packages/bitcore-node/src/services/p2p/index.ts]: should be `const`

File: packages/bitcore-node/src/services/p2p/index.ts [hunk 2]
```diff
@@ -45,21 +51,17 @@ export interface P2pService<Block, Transaction> {
...
-  block: B,
-  transactions: T[],
+  block: B;
+  transactions: T[];
 };
 
-
 export enum P2pEvents {
-  SYNC_COMPLETE = 'SYNC_COMPLETE',
+  SYNC_COMPLETE = 'SYNC_COMPLETE'
```
Comment (illegalprime) [inline packages/bitcore-node/src/services/p2p/index.ts:65]: nit: I like the trailing commas `,` they make it easier to add items.

File: packages/bitcore-node/src/services/p2p/index.ts [hunk 3]
```diff
@@ -84,68 +86,86 @@ export class P2pRunner {
     this.events = new EventEmitter();
   }
 
+  async wireupBlockStream(syncer, parent?) {
```
Comment (illegalprime) [inline packages/bitcore-node/src/services/p2p/index.ts]: these args need types:
```typescript
async wireupBlockStream(syncer: { add: (string) => void; }, parent?: { height: number; chain: string }): Observable<_>
```

File: packages/bitcore-node/src/services/p2p/index.ts [hunk 4]
```diff
@@ -84,68 +86,86 @@ export class P2pRunner {
...
+              network: this.network
+            });
+          }
+          return pair;
+        })
+      )
+      .pipe(share());
+  }
+
+  async wireupTxStream() {
```
Comment (illegalprime) [inline packages/bitcore-node/src/services/p2p/index.ts]: needs return type

File: packages/bitcore-node/src/services/p2p/index.ts (hunk 3 above)
Comment (illegalprime) [reply packages/bitcore-node/src/services/p2p/index.ts]: Also maybe make these methods `private`? It doesn't make sense to me for a user to be able to wire up something only partially, and to have to remember to do both, and maybe do it twice accidentally.

File: packages/bitcore-node/src/services/p2p/index.ts [hunk 5]
```diff
@@ -207,101 +226,89 @@ export class P2pRunner {
...
+  get({ chain }: Chain): Class<StandardP2p> {
+    if (P2PClasses[chain]) {
+      return P2PClasses[chain];
     }
     throw new Error(`Chain ${chain} doesn't have a P2P Worker registered`);
   }
 
-  register(chain: string, service: P2pRunner) {
-    this.services[chain] = service;
+  register(chain: string, service: Class<StandardP2p>) {
```
Comment (illegalprime) [inline packages/bitcore-node/src/services/p2p/index.ts:264]: Even if this compiles its semantics aren't quite right, `P2PClasses` has a type of:
```typescript
{ [key in keyof typeof SupportedChainSet]: Class<StandardP2p> }
```
So you cannot add `chain: string` to it.

File: packages/bitcore-node/src/services/p2p/index.ts [hunk 6]
```diff
@@ -207,101 +226,89 @@ export class P2pRunner {
...
-
-  constructor() {
-    this.services = {};
-  }
-
-  get({ chain }: Chain): P2pRunner {
-    if (this.services[chain]) {
-      return this.services[chain];
+export class P2pProxy implements CSP.Provider<Class<StandardP2p>> {
+  get({ chain }: Chain): Class<StandardP2p> {
```
Comment (illegalprime) [inline packages/bitcore-node/src/services/p2p/index.ts:257]: Maybe this should also be `chain: string` like `register`?

File: packages/bitcore-node/src/services/p2p/index.ts [hunk 7]
```diff
@@ -207,101 +226,89 @@ export class P2pRunner {
...
             blocks: BlockModel,
             transactions: TransactionModel,
-            config: p2pServiceConfig,
+            config: p2pServiceConfig
           });
-          this.register(chain, runner);
 
           // get ready to start the service
-          p2pServices.push(() => runner.start());
+          p2pServices.push(runner.start());
```
Comment (illegalprime) [inline packages/bitcore-node/src/services/p2p/index.ts:312]: This starts them immediately on this line, is that what you want?

File: packages/bitcore-node/src/types/Class.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+type Class<T> = { new (...args: any[]): T };
```
Comment (illegalprime) [inline packages/bitcore-node/src/types/Class.ts:1]: :+1: Maybe `Constructable`?

**Review by illegalprime — CHANGES_REQUESTED** [review]
Looks pretty good, just got a few small changes.

File: packages/bitcore-node/src/services/p2p/index.ts (hunk 2 above)
Comment (micahriggan) [reply packages/bitcore-node/src/services/p2p/index.ts:65]: I don't think our linter file has that, or mine isn't respecting it

**Review by illegalprime — APPROVED** [review]

**Pull request merged**